supporter-product-data/src/main/scala/com/gu/services/S3Service.scala

```diff
@@ -0,0 +1,49 @@
+package com.gu.services
+
+import com.amazonaws.regions.Regions
+import com.amazonaws.services.s3.AmazonS3ClientBuilder
+import com.amazonaws.services.s3.model.{GetObjectRequest, ObjectMetadata}
+import com.amazonaws.services.s3.transfer.TransferManagerBuilder
+import com.gu.aws.CredentialsProvider
+import com.gu.model.Stage
+import com.gu.monitoring.SafeLogger
+
+import java.io.InputStream
+import java.util.UUID
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Future, blocking}
+
+object S3Service {
+  val s3Client = AmazonS3ClientBuilder.standard
+    .withRegion(Regions.EU_WEST_1)
+    .withCredentials(CredentialsProvider)
+    .build
+  val transferManager = TransferManagerBuilder.standard
+    .withS3Client(s3Client)
+    .build
+
+  def bucketName(stage: Stage) = s"supporter-product-data-export-${stage.value.toLowerCase}"
+
+  def streamToS3(stage: Stage, filename: String, inputStream: InputStream, length: Long) = {
+    // Specify server-side encryption.
+    val objectMetadata = new ObjectMetadata()
+    objectMetadata.setContentLength(length)
+    objectMetadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION)
+
+    val upload = transferManager.upload(
+      bucketName(stage),
+      filename,
+      inputStream,
+      objectMetadata
+    )
+
+    Future(
+      blocking(upload.waitForUploadResult())
```

Comment:
`blocking` doesn't always do anything useful depending on the execution context, it can hang if the number of threads can't expand. Why not addProgressListener and listen for a TRANSFER_COMPLETED_EVENT in a similar way to our AwsAsyncHandler class?

Reply:
Good suggestion, have implemented